### PR TITLE
Focus widget

### PR DESCRIPTION
**Pull request by kmerz**
## Motivation and Context
The user wants to set a single widget in focus by removing all other widgets temporarily and see the focused widget in full screen.
 
## Description
- Add a new CurrentViewState[Store/Actions] which will store the widget in focus
- Render the focused widget instead of the WidgetGrid
- Unset focused Widget before going into big display mode
- Adjust css styling

## Screenshots (if appropriate):
fullscreen-widget

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: graylog2-web-interface/src/views/stores/CurrentViewStateStore.ts [hunk 1]
```diff
@@ -47,11 +48,13 @@ export const CurrentViewStateActions: CurrentViewStateActionsType = singletonAct
     titles: { asyncResult: true },
     widgets: { asyncResult: true },
     widgetPositions: { asyncResult: true },
+    focusWidget: { asyncResult: true },
   }),
 );
 
 type CurrentViewStateStoreState = {
   activeQuery: string;
+  focusedWidget: string | undefined | null,
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/views/stores/CurrentViewStateStore.ts:57]: Do we need to put this in the `CurrentViewStateStore`? I would suggest creating a separate context for this instead, which provides the current state and the actions to modify it.

**Review by dennisoelkers — CHANGES_REQUESTED** [review]
Works! Unfortunately it creates a scroll bar:

image

File: graylog2-web-interface/src/views/stores/CurrentViewStateStore.ts (hunk 1 above)
Comment (linuspahl) [reply graylog2-web-interface/src/views/stores/CurrentViewStateStore.ts:57]: @kmerz We could use the `SearchPageLayoutContext` for this.

**Review by linuspahl — CHANGES_REQUESTED** [review]
Looks good, some thoughts and findings:

- When I open a dashboard I sometimes receive the error `Cannot read property 'id' of undefined` (`Query` - `_renderWidgetGrid`)
- We should make clear that the focus is only a temporary state. I can imagine that a user will not directly understand that this action behaves differently than e.g. the `Stretch width` action.
image
We could show a label (when focused) for the new action. In this case I would move the `Focus / Unfocus widget` before the dropdown icon (in both states). We should also hide (or disable) the `Stretch widget` action when the focus is active.

File: graylog2-web-interface/src/views/components/SearchResult.tsx [hunk 2]
```diff
@@ -56,6 +67,7 @@ type Props = {
 
 const SearchResult = React.memo(({ queryId, searches, viewState }: Props) => {
   const fieldTypes = useContext(FieldTypesContext);
+  const focusedWidget = useContext(WidgetFocusContext);
```
Comment (linuspahl) [inline graylog2-web-interface/src/views/components/SearchResult.tsx:70]: We need to use ` const { focusedWidget } = useContext(WidgetFocusContext);` here as well, right now `focusedWidget`  is always true.

**Review by linuspahl — COMMENTED** [review]
As discussed we do not have to adjust the widget actions in the "is focused" view right now, but we should disable actions like stretch and duplicate.

**Review by linuspahl — APPROVED** [review]
Looks like the duplicate action is still enabled. I am still approving the PR so my change request will not block the merge next week.

File: graylog2-web-interface/src/views/components/Query.test.jsx [hunk 3]
```diff
@@ -57,14 +64,14 @@ describe('Query', () => {
...
-      <Query results={results}
-             widgetMapping={widgetMapping}
-             widgets={widgets}
-             onToggleMessages={() => {}}
-             queryId="someQueryId"
-             showMessages
-             allFields={Immutable.List()}
-             fields={Immutable.List()} />
+      <SUT results={results}
+           widgetMapping={widgetMapping}
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/views/components/Query.test.jsx:68]: Can we move all of the props which are constantly reused to the `SUT` definition? This way we can separate the boilerplate and override in tests what's actually different, which makes it way easier to understand.

File: graylog2-web-interface/src/views/components/Query.tsx [hunk 4]
```diff
@@ -81,6 +87,27 @@ const _renderWidgetGrid = (widgetDefs, widgetMapping, results, positions, queryI
     }
   });
 
+  if (focusedWidget) {
```
Comment (dennisoelkers) [inline graylog2-web-interface/src/views/components/Query.tsx:90]: I think we should move this to a separate function itself (`renderFocussedWidget`?) and do a conditional in the `Query` function, e.g. `return focussedWidget ? renderFocussedWidget(...) : renderWidgetGrid(...);` - this makes it easy to understand.

**Review by dennisoelkers — CHANGES_REQUESTED** [review]
LGTM, just a few remarks inline.

**Review by dennisoelkers — CHANGES_REQUESTED** [review]
Looks better already, two smaller comments inline.

**Review by dennisoelkers — APPROVED** [review]
Works fine, great work! :tada:

**Pull request merged**